### PR TITLE
Application v2 - DO NOT MERGE

### PR DESCRIPTION
**Pull request by judy2k**

### CODE REVIEW

File: Nexmo.Api/Application.cs [hunk 1]
```diff
@@ -61,10 +62,10 @@ public class KeyPair
 
     public class VoiceHooks
     {
-        public Webhook[] webhooks { get; set; }
```
Comment (judy2k) [inline Nexmo.Api/Application.cs]: I hate to say it, but 'webhook' is generally one word these days, so the previous version was probably more correct.

https://en.wikipedia.org/wiki/Webhook

File: Nexmo.Api/Application.cs [hunk 2]
```diff
@@ -98,6 +99,7 @@ public class ApplicationList
     public class ApplicationListResponse : PaginatedResponse<ApplicationList>
     {}
 
+    [Obsolete("This version of the API has been depricated", true)]
```
Comment (judy2k) [inline Nexmo.Api/Application.cs]: "deprecated"

File: Nexmo.Api/ApplicationV2.cs [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+        /// <summary>
+        /// The name of your application. Required.
+        /// </summary>
+        [JsonProperty("name")]
+        public string Name { get; set; }
+        /// <summary>
+        /// The private and public keys to be used with the Application. 
+        /// </summary>
+        [JsonProperty("keys")]
+        public Keys Keyes { get; set; }
```
Comment (judy2k) [inline Nexmo.Api/ApplicationV2.cs]: Typo.

File: Nexmo.Api/ApplicationV2.cs [hunk 4]
```diff
@@ -0,0 +1,263 @@
...
+        public Capabilities Capabilities { get; set; }
+    }
+
+    public class Keys
+    {
+        /// <summary>
+        /// The public key used to validate the jwt in a https://en.wikipedia.org/wiki/JSON_Web_Token.
+        /// </summary>
+        [JsonProperty("private_key")]
+        public string Private_key { get; set; }
```
Comment (judy2k) [inline Nexmo.Api/ApplicationV2.cs]: This combination of snake case and camel case seems odd to me - would it be better as `PrivateKey`?

File: Nexmo.Api/Client/Client.cs [hunk 5]
```diff
@@ -43,7 +44,9 @@ private void PropagateCredentials()
 
         public ClientMethods.Account Account { get; private set; }
         public ClientMethods.ApiSecret ApiSecret { get; private set; }
+        [Obsolete("This version of the API has been depricated", true)]
         public ClientMethods.Application Application { get; private set; }
+        public ClientMethods.ApplicationV2 ApplicationV2 { get; private set; }
```
Comment (judy2k) [inline Nexmo.Api/Client/Client.cs]: It looks like you're no longer initialising `Application`, but you're leaving the accessor there. I assume this means that if the user tries to access `ClientMethods.Application` they'll get `null`?

This seems like a really bad idea. I'd recommend: _Either_ deprecate `Application` and leave it working, _or_ remove it entirely, so the user's code won't compile if they're accessing it.

File: Nexmo.Api/ApplicationV2.cs [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+        public static AppResponse Get(string appId, Credentials credentials = null)
+        {
+            // This is a dirty hack for the GET requests on Application V2 to work
+            // until the fix on the API level is implemented.
+#if (!NETSTANDARD1_6)
+            {
+                string url = ApiRequest.GetBaseUriFor(typeof(ApplicationV2), $"/v2/applications/{appId}").ToString();
+                string result;
+                var authBytes = Encoding.UTF8.GetBytes(credentials.ApiKey + ":" + credentials.ApiSecret);
+                using (WebClient client = new WebClient())
```
Comment (judy2k) [inline Nexmo.Api/ApplicationV2.cs]: 🤷‍♂ 😄

File: Nexmo.Api/ApplicationV2.cs [hunk 7]
```diff
@@ -0,0 +1,263 @@
...
+            //    { "page", page.ToString()}
+            //};
+
+            //var json = VersionedApiRequest.DoRequest("GET", ApiRequest.GetBaseUriFor(typeof(ApplicationV2), "/v2/applications"), searchFilter, credentials);
+
+            //var response = JsonConvert.DeserializeObject<AppListResponse>(json.JsonResponse);
+            //return response._embedded.Applications;
+        }
+
+        public static List<AppResponse> List()
```
Comment (judy2k) [inline Nexmo.Api/ApplicationV2.cs]: Are you missing some docs for this method?

File: Nexmo.Api/ApplicationV2.cs (hunk 4 above)
Comment (RabebOthmani) [reply Nexmo.Api/ApplicationV2.cs]: Absolutely, I was thinking about the JSON property...
https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/capitalization-conventions

File: Nexmo.Api/ApplicationV2.cs (hunk 3 above)
Comment (RabebOthmani) [reply Nexmo.Api/ApplicationV2.cs]: 🤦

File: Nexmo.Api/Application.cs (hunk 2 above)
Comment (RabebOthmani) [reply Nexmo.Api/Application.cs]: 🤦

File: Nexmo.Api/ApplicationV2.cs (hunk 7 above)
Comment (RabebOthmani) [reply Nexmo.Api/ApplicationV2.cs]: Nope, meant to be like this :
```
public static List<AppResponse> List()
        {
            return List(10, 0);
        }
```

File: Nexmo.Api/Client/Client.cs (hunk 5 above)
Comment (RabebOthmani) [reply Nexmo.Api/Client/Client.cs]: When a class is marked as obsolete, you can't instantiate it anymore, it throws a compiling error. I'll completely remove it from Client. Leaving the class becomes irrelevant also but i'm not sure if completely removing the code is the best decision

Comment (judy2k) [reply Nexmo.Api/Client/Client.cs]: 👍

File: Nexmo.Api/Client/Application.cs [hunk 8]
```diff
@@ -1,8 +1,10 @@
-﻿using System.Collections.Generic;
+﻿using System;
+using System.Collections.Generic;
 using Nexmo.Api.Request;
 
 namespace Nexmo.Api.ClientMethods
 {
+    [Obsolete("This version of the API has been deprecated", true)]
```
Comment (judy2k) [inline Nexmo.Api/Client/Application.cs]: Should probably say " ... use ApplicationV2 instead"

File: Nexmo.Api/Application.cs (hunk 1 above)
Comment (judy2k) [reply Nexmo.Api/Application.cs]: OK, just did a bit more research, and Microsoft seem to consistently use WebHooks, so this is fine as it is. Microsoft! 🤦‍♂

**Pull request merged**